src/server/tiered_storage.cc

```diff
@@ -355,6 +356,68 @@ void TieredStorage::Free(PrimeIterator it, DbTableStats* stats) {
   stats->tiered_size -= len;
 }
 
+size_t TieredStorage::GetBinSize(size_t blob_len) {
+  CHECK(blob_len <= kMaxSmallBin);
+  unsigned bin_index = SmallToBin(blob_len);
+  return kSmallBins[bin_index];
+}
+
+void TieredStorage::Defrag(DbIndex db_index) {
+  // start scanning from a random position in the ref_cnt_ table.
+  auto refcnt_it = std::next(std::begin(page_refcnt_), rand() % page_refcnt_.size());
+
+  for (unsigned int i = 0; i < num_pages_to_scan_; ++i) {
```

Comment:
I dont think you need to limit the number of entries we check in the map but instead limit the number of read operations we do.
i.e you can go over the entier map non of the pages will be below the threshold and you will finish very fast as there is no operation if no defrag is needed. 
while you will do want to limit the number of reads from disk every time this defrag function is called

Reply:
that's certainly another option. It is hard to say which one is better... the defragmentation happens when the CPU is idle.

Reply:
> I dont think you need to limit the number of entries we check in the map but instead limit the number of read operations we do. i.e you can go over the entier map non of the pages will be below the threshold and you will finish very fast as there is no operation if no defrag is needed. while you will do want to limit the number of reads from disk every time this defrag function is called

the other question i had on limiting the number of reads is that if all the pages have good utlization and do not need defragmentation, would we keep iterating forever as the number of read is always zero? we do not need to read any page to calculate utilization as we already know the reference count and bin size.